### PR TITLE
Enzyme tests take 2

### PR DESCRIPTION
**Pull request by mccarthy417nahant**
merged with Carrier Wave branch - enzyme tests are finally working

### CODE REVIEW

File: app/controllers/api/v1/reviews_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  end
+  
+  def create
+    review = Review.new(body: params[:body], rating: params[:rating], user_id: params[:user_id], bike_id: params[:bike_id])
+    if review.save
+      render json: review
+    else
+      render json: review.errors.full_messages, status: :unprocessable_entity
+    end
+  end
```
Comment (nwalberts) [inline app/controllers/api/v1/reviews_controller.rb]: You need to be using strong params for the creation of your new Review

File: app/controllers/api/v1/reviews_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+      @bike = Bike.find(params['id'])
+      @reviews = @bike.reviews
+
+    render json: @reviews
+  end
+
+  def index
+    @reviews = Review.all
+    render json: @reviews
+  end
```
Comment (nwalberts) [inline app/controllers/api/v1/reviews_controller.rb]: Are you even using this endpoint?

File: app/controllers/api/v1/reviews_controller.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class Api::V1::ReviewsController < ApiController
+  def show
+      @bike = Bike.find(params['id'])
+      @reviews = @bike.reviews
+
+    render json: @reviews
+  end
```
Comment (nwalberts) [inline app/controllers/api/v1/reviews_controller.rb]: Update indentation

File: app/controllers/api/v1/bikes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class Api::V1::BikesController < ApiController
+  def show
+    render json: Bike.find(params[:id])
+  end
+
+end
```
Comment (nwalberts) [inline app/controllers/api/v1/bikes_controller.rb]: Make your whitespace consistent. Either add above def show or remove below def show

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    @users = User.all
+  end
+  def show
+    @user = User.find(params[:id])
+  end
+
+  def new
+    @user = User.new
+  end
+
```
Comment (nwalberts) [inline app/controllers/users_controller.rb]: Do you even use this endpoint?

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  def update
+    @user = User.find(params[:id])
+    if @user.update_attributes(account_update_params)
+      redirect_to user_path
+    end
+  end
```
Comment (nwalberts) [inline app/controllers/users_controller.rb]: You should have a condition if `update_attributes` does NOT work

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+  private
+  #
+  # def sign_up_params
+  #   params.require(:user).permit(:email, :encrypted_password, :profile_photo)
+  # end
+
```
Comment (nwalberts) [inline app/controllers/users_controller.rb]: remove unused code/comments

File: app/javascript/react/components/BikeShow.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      <h2>{props.model}</h2>
+      <h3>{props.make}</h3>
+      <h3>{props.year}</h3>
+      <h3>{props.code}</h3>
+
+      <h3>Reviews!</h3>
+    </div>
+  )
+
+}
```
Comment (nwalberts) [inline app/javascript/react/components/BikeShow.js]: remove extra whitespace

File: app/javascript/react/components/ReviewTile.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+const ReviewTile = (props) => {
+  return (
+    <div key={props.id}>
+      <h6> {props.email}</h6>
+      <p className="rating"> {props.body} - {props.rating} stars</p>
+    </div>
+  )
+
+}
```
Comment (nwalberts) [inline app/javascript/react/components/ReviewTile.js]: remove extra whitespace

File: app/javascript/react/containers/BikeShowContainer.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    fetch(`/api/v1/bikes/${this.props.params.id}`)
+      .then(response => response.json())
+      .then(body => {
+        this.setState({
+          bike: body.bike,
+          reviews: body.bike.reviews
+        });
+    })
+  }
+
```
Comment (nwalberts) [inline app/javascript/react/containers/BikeShowContainer.js]: You need error handling in this fetch request!

File: app/javascript/react/containers/BikeShowContainer.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+  addSubmission(submission) {
+
+    this.setState({reviews: this.state.reviews.concat(submission)})
```
Comment (nwalberts) [inline app/javascript/react/containers/BikeShowContainer.js]: you should _not_ be adding to your reviews state until AFTER your post request is complete, and when you do you should use the new review returned from your backend

File: app/javascript/react/containers/BikeShowContainer.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    fetch('/api/v1/reviews', {
+      credentials: 'same-origin',
+      method: "post",
+      body: JSON.stringify(submission),
+      headers: { 'Content-Type': 'application/json' }
+    })
+
```
Comment (nwalberts) [inline app/javascript/react/containers/BikeShowContainer.js]: remove whitespace

File: app/javascript/react/containers/BikeShowContainer.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        error = new Error(errorMessage);
+        throw(error);
+      }
+    })
+    .then(response => response.json())
+    .then(body => {
+      alert("Success!")
+      // this.setState({reviews: this.state.reviews.concat(body.reviews)})
+    })
+    .catch(error => console.error('Error:', error));
```
Comment (nwalberts) [inline app/javascript/react/containers/BikeShowContainer.js]: you should be returned a newly created review from your backend and adding it to state. 
also remove comment

File: app/javascript/react/containers/BikeShowContainer.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    })
+    return(
+      <div>
+        <BikeShow
+          id={this.state.bike.id}
+          make={this.state.bike.make}
+          model={this.state.bike.model}
+          year={this.state.bike.year}
+          code={this.state.bike.code}
+          // image_url={this.state.bike.profile_photo.url}
```
Comment (nwalberts) [inline app/javascript/react/containers/BikeShowContainer.js]: remove old comments

File: app/javascript/react/containers/FormContainer.js [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+componentDidMount() {
+  fetch(`/api/v1/users`)
+    .then(response => response.json())
+    .then(body => {
+      this.setState({
+        current_user: body
+      })
+
+    })
+}
```
Comment (nwalberts) [inline app/javascript/react/containers/FormContainer.js]: error handling needed! 
remove whitepace

File: app/javascript/react/containers/FormContainer.js [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+handleDescriptionChange(event) {
+  this.setState({body: event.target.value})
+}
+handleRatingChange(event) {
+  this.setState({rating: event.target.value})
+}
```
Comment (nwalberts) [inline app/javascript/react/containers/FormContainer.js]: add whitespace between method definitions

File: app/javascript/react/containers/FormContainer.js [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        name="description"
+      />
+      <TextField
+        label = "Rating of Bike"
+        content = {this.state.rating}
+        handleChange = {this.handleRatingChange}
+        name="rating"
+      />
+    <input type="submit" className="button" value="Submit "/>
+    </form>
```
Comment (nwalberts) [inline app/javascript/react/containers/FormContainer.js]: fix indentation on line 62

File: app/models/bike.rb [hunk 18]
```diff
@@ -2,4 +2,9 @@ class Bike < ApplicationRecord
   validates :make, presence:true
   validates :model, presence:true
   validates :year, presence: true, numericality: true, length: {is: 4}
+
+  belongs_to :user
+  has_many :reviews
+
+  mount_uploader :profile_photo, ProfilePhotoUploader
```
Comment (nwalberts) [inline app/models/bike.rb]: There should NOT be a profile photo uploader here

File: app/serializers/review_serializer.rb [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+  attributes :id, :body, :rating, :user_email, :created_at, :user_id
+
+  def user_email
+    "#{object.user.email}"
+  end
+
+  # def user
+  #   "#{object.user.id}"
+  # end
+
```
Comment (nwalberts) [inline app/serializers/review_serializer.rb]: remove unused code

File: app/uploaders/profile_photo_uploader.rb [hunk 20]
```diff
@@ -0,0 +1,47 @@
+class ProfilePhotoUploader < CarrierWave::Uploader::Base
+  # Include RMagick or MiniMagick support:
+  # include CarrierWave::RMagick
+  # include CarrierWave::MiniMagick
+
+  # Choose what kind of storage to use for this uploader:
+  storage :file
+  # storage :fog
```
Comment (nwalberts) [inline app/uploaders/profile_photo_uploader.rb]: you should be using fog a soon as your can, which is in the Cloud storage article

File: db/migrate/20181015195242_create_reviews.rb [hunk 21]
```diff
@@ -0,0 +1,9 @@
+class CreateReviews < ActiveRecord::Migration[5.2]
+  def change
+    create_table :reviews do |t|
+      t.belongs_to :bike, null: false
+      t.belongs_to :user, null: false
+      t.timestamps
+    end
+  end
+end
```
Comment (nwalberts) [inline db/migrate/20181015195242_create_reviews.rb]: you should straight up just update this migration with all of the other migrations you made in this PR

File: db/migrate/20181015195822_add_belongs_to_bikes.rb [hunk 22]
```diff
@@ -0,0 +1,5 @@
+class AddBelongsToBikes < ActiveRecord::Migration[5.2]
+  def change
+    add_reference :bikes, :user, index: true
+  end
```
Comment (nwalberts) [inline db/migrate/20181015195822_add_belongs_to_bikes.rb]: you should not be adding references, just add a column of user_id to a bike

File: db/migrate/20181016191937_add_column_to_reviews.rb [hunk 23]
```diff
@@ -0,0 +1,5 @@
+class AddColumnToReviews < ActiveRecord::Migration[5.2]
+  def change
+    add_column :reviews, :body, :string
+  end
```
Comment (nwalberts) [inline db/migrate/20181016191937_add_column_to_reviews.rb]: delete this migration and edit your create migration

File: spec/features/user_checks_profile_spec.rb [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+    fill_in 'Email', with: "john@example.com"
+    fill_in 'Password', with: 'password'
+    fill_in 'Password confirmation', with: 'password'
+    click_button 'Sign up'
+    expect(page).to have_content('Welcome! You have signed up successfully.')
+    expect(page).to have_content('Sign Out')
+    save_and_open_page
+    click_link 'Profile'
+    expect(page).to have_content("john@example.com")
+  end
```
Comment (nwalberts) [inline spec/features/user_checks_profile_spec.rb]: I dont see a test for the profile photo. The reading has an example test

File: spec/javascript/container/BikeShowContainer.js [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+    params ={ {id: bikeData.bike.id} }
+      />
+    );
+  });
+
+  afterEach(fetchMock.restore)
+
+  it('should render bike model in h2 tags', (done) =>{
+    setTimeout(() => {
+      // console.log(wrapper.debug())
```
Comment (nwalberts) [inline spec/javascript/container/BikeShowContainer.js]: remove comments

File: spec/javascript/container/BikeShowContainer.js [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+      expect(wrapper.containsMatchingElement(<h2>Electra Glide</h2>)).toBeTruthy()
```
Comment (nwalberts) [inline spec/javascript/container/BikeShowContainer.js]: you should not be usiing `toBeTruthy` but insteead .toMatch, or any of the other matchers used in the readings

File: spec/javascript/container/BikeShowContainer.js [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+    setTimeout(() => {
+      // console.log(wrapper.debug())
+      rating = wrapper.find('.rating')
+      // console.log(rating.debug())
+      expect(rating.text()).toEqual(" Itza okay. - 6 stars")
+    done()
+    }, 0)
+  });
+
+});
```
Comment (nwalberts) [inline spec/javascript/container/BikeShowContainer.js]: REMOVE OLD COMMENTS

Comment (nwalberts) [inline spec/javascript/container/BikeShowContainer.js]: Please review the readings for more appropriate enzyme tests. If you have questions, add them to the QQ

File: app/controllers/api/v1/reviews_controller.rb (hunk 1 above)
Comment (mccarthy417nahant) [reply app/controllers/api/v1/reviews_controller.rb]: done

File: app/controllers/api/v1/reviews_controller.rb (hunk 2 above)
Comment (mccarthy417nahant) [reply app/controllers/api/v1/reviews_controller.rb]: using for post

File: app/controllers/api/v1/reviews_controller.rb (hunk 3 above)
Comment (mccarthy417nahant) [reply app/controllers/api/v1/reviews_controller.rb]: done

File: app/controllers/api/v1/bikes_controller.rb (hunk 4 above)
Comment (mccarthy417nahant) [reply app/controllers/api/v1/bikes_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (mccarthy417nahant) [reply app/controllers/users_controller.rb]: removed

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (mccarthy417nahant) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 7 above)
Comment (mccarthy417nahant) [reply app/controllers/users_controller.rb]: done

File: app/javascript/react/components/BikeShow.js (hunk 8 above)
Comment (mccarthy417nahant) [reply app/javascript/react/components/BikeShow.js]: done

File: app/javascript/react/components/ReviewTile.js (hunk 9 above)
Comment (mccarthy417nahant) [reply app/javascript/react/components/ReviewTile.js]: done

File: app/javascript/react/containers/BikeShowContainer.js (hunk 10 above)
Comment (mccarthy417nahant) [reply app/javascript/react/containers/BikeShowContainer.js]: done

File: app/javascript/react/containers/BikeShowContainer.js (hunk 11 above)
Comment (mccarthy417nahant) [reply app/javascript/react/containers/BikeShowContainer.js]: done

File: app/javascript/react/containers/BikeShowContainer.js (hunk 12 above)
Comment (mccarthy417nahant) [reply app/javascript/react/containers/BikeShowContainer.js]: done

File: app/javascript/react/containers/BikeShowContainer.js (hunk 14 above)
Comment (mccarthy417nahant) [reply app/javascript/react/containers/BikeShowContainer.js]: done

File: app/javascript/react/containers/FormContainer.js (hunk 15 above)
Comment (mccarthy417nahant) [reply app/javascript/react/containers/FormContainer.js]: done

File: app/javascript/react/containers/FormContainer.js (hunk 16 above)
Comment (mccarthy417nahant) [reply app/javascript/react/containers/FormContainer.js]: done

File: app/javascript/react/containers/FormContainer.js (hunk 17 above)
Comment (mccarthy417nahant) [reply app/javascript/react/containers/FormContainer.js]: done

File: app/models/bike.rb (hunk 18 above)
Comment (mccarthy417nahant) [reply app/models/bike.rb]: removed

File: app/serializers/review_serializer.rb (hunk 19 above)
Comment (mccarthy417nahant) [reply app/serializers/review_serializer.rb]: done

File: app/uploaders/profile_photo_uploader.rb (hunk 20 above)
Comment (mccarthy417nahant) [reply app/uploaders/profile_photo_uploader.rb]: done

File: db/migrate/20181015195242_create_reviews.rb (hunk 21 above)
Comment (mccarthy417nahant) [reply db/migrate/20181015195242_create_reviews.rb]: done

File: db/migrate/20181015195822_add_belongs_to_bikes.rb (hunk 22 above)
Comment (mccarthy417nahant) [reply db/migrate/20181015195822_add_belongs_to_bikes.rb]: done

File: db/migrate/20181016191937_add_column_to_reviews.rb (hunk 23 above)
Comment (mccarthy417nahant) [reply db/migrate/20181016191937_add_column_to_reviews.rb]: done

File: spec/javascript/container/BikeShowContainer.js (hunk 25 above)
Comment (mccarthy417nahant) [reply spec/javascript/container/BikeShowContainer.js]: done

File: app/uploaders/profile_photo_uploader.rb (hunk 20 above)
Comment (mccarthy417nahant) [reply app/uploaders/profile_photo_uploader.rb]: done

**Comment by d-mui22** [comment]
Actually do use this. This has the syntax changes Nick specified in the comments. Good Luck. sorry

**Pull request closed without merging**